video: Merge driver stream start/stop APIs

ngphibang opened the pull request:
The video_stream_start/stop() APIs are counter-symetric and have the same function signature. Also, the implementation logic for those APIs is generally the same. Merge these **two drivers API** to save memory and code lines (e.g., **number of code lines are saved by nearly a half**)

**For the sake of simplicity, still keep the user APIs to keep backward compatibility with downstream applications.**

While at it:

- Fix a related bug in video_emul_rx
- Fix video related doc in migration guide 4.1

ngphibang committed:
drivers: video: emul_rx: Fix flush function

When stopping, the framework calls driver's stop callback then driver's
flush callback. Hence, driver's flush callback does not need to call
stop callback again.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
doc: release: 4.1: Move video_buffer_alloc to video section

Move video_buffer_alloc to the video section.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -583,49 +574,34 @@ static inline int video_flush(const struct device *dev, enum video_endpoint_id e
+ * @retval 0 on success, otherwise a negative errno code.
  */
-static inline int video_stream_start(const struct device *dev)
+static inline int video_stream_enable(const struct device *dev, bool enable)
```
`video_stream_enable(dev, false)` is less obvious at glance than `video_stream_stop(dev)`.
There could be a `video_stream_disable(dev, false)`.

`video_flush(dev, ep, false)` will still flush even if the last argument is `false`.

Something like `video_set_stream(dev, false)` would allow to guess that the device's stream status becomes `false` (not streaming).

Maybe `video_stream_enable()` is still a better compromise?

josuah reviewed (commented):
I do like the change!

A non-blocking suggestion to use a more foolproof name. Not picky on what as long as it avoids quidproquos.

Good catch on the emul-rx!

loicpoulain commented:
Make sense for the driver API, but for simplicity I think we could keep the user API stream start and stop, and simply have a macro or inline function calling enable with the right parameter. (no cost in term of memory).

ngphibang commented:
> Make sense for the driver API, but for simplicity I think we could keep the user API stream start and stop, and simply have a macro or inline function calling enable with the right parameter. (no cost in term of memory).

Yes, it's a good idea. Technically, it will be no much different. The advantage of keeping the user API unchanged is to not break downstream applications (?). Personnally, I prefer a complete change but it seems in this case, it's better to keep the downstream compatibility. Will do.

josuah commented:
There is a `__deprecated` flag (for `static inline video_stream_start()`) or `__DEPRECATED_MACRO` (for `#define video_stream_start()`) supported in Zephyr if that helps:
https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated

Not sure if the whole communication process applies for areas that are not "maintained" (which requires an official maintainer and a few other criterions). Still good as a guideline, i.e. remove the deprecated call in `4.3`?

ngphibang committed:
video: Merge video_stream_start/stop driver APIs

The video_stream_start/stop() APIs are counter-symetric and have
the same function signature. Also, the implementation logic for
those driver APIs is generally the same. Merge them to save memory
and code lines.

For the sake of simplicity, still keep the user APIs to preserve
backward compatibility with downstream applications.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
Force-push:

- Just merge drivers API, keep user APIs unchanged (i.e., all the application / sample codes keep unchanged) as discussed
- Renamed video_stream_enable() to video_set_stream() as discussed
- Fixed a small CI issue

JarmouniA commented:
> but for simplicity I think we could keep the user API stream start and stop, and simply have a macro or inline function calling enable with the right parameter. (no cost in term of memory).

This makes more since & avoids breakage. SPI does a similar thing https://github.com/zephyrproject-rtos/zephyr/blob/eb8fe504f7a074b3d3b1e0f58ddcd7750ec4866c/include/zephyr/drivers/spi.h#L829

Plus the existing start/stop API is a lot clearer for users than the proposed change. I don't see any value in changing things on the app side.

JarmouniA commented on doc/releases/release-notes-4.1.rst:
```diff
@@ -77,6 +77,9 @@ Removed APIs and options
   and only supported 8-bit depth to :c:func:`video_bits_per_pixel()` returning
   the *bit* count and supporting any color depth.
 
+* The ``video_stream_start()`` and ``video_stream_stop()`` APIs have been removed
```
```suggestion
* The ``video_stream_start()`` and ``video_stream_stop()`` driver APIs have been removed
```
